### DISCUSSION TITLE
How accurate is time keeping with the 2-series?

### DISCUSSION THREAD

**Discussion by InstantMuffin** [SpenceKonde/megaTinyCore#792] [Library discussion]
Hello,

Long story short. I want to create a datalogger in as little time as possible.
The device will use a 2-series ATTiny, and operate on a battery. Every few seconds I'm going to do stuff to count up a number of events, and I want to log that count to the eeprom for every hour. I don't have an external eeprom either so my logging is pretty limited that way, needing 1 byte per entry I am left with about 240 hours or 10 days worth of data.
I would say a maximum of 10 minutes drift after those 10 days would be acceptable.

Reading about oscillator accuracy I see there are two factors:
1) There is a fixed drift which can be calibrated for.
2) There is some random (how random is it really since that might introduce incremental shifting as well) inaccuracies which I would debate on whether or not these cancel each other out over time.

Has anyone of you done something similar without using or rather having the time to order and use a dedicated RTC IC and can share his or her experiences?
Is it possible without calibration, only with calibration, or is this not feasible at all without using a dedicated RTC for time keeping in this specific scenario?

Thank you for your time and answers!

**Comment by ObviousInRetrospect** [comment]
Your target (10 minutes in 10 days) is 1000000*(600/864000) = 694 PPM or 0.069%. This is trivially achievable with a watch crystal and going to be hard to hit without one.

Assembling blind (calculating the correct value but not actually tuning) I tend to get about +/-75 PPM on a watch crystal (or +/- 150 when being dumb and using horribly wrong values). With a bit of trial and error I can get it down to +/- a couple PPM (IF TEMPERATURES ARE STABLE). There is an adjustment on board to bring it down to whatever fraction of a PPM you are away from an integer. If you want precision, buy some +/- 0.1pf caps (8.2, 8.5, 9, 9.1, 10, 11 for a 6pf crystal) and/or layout a pair of caps in parallel to allow the addition of a 2nd 0-2pf to tune.  Done right you can get +/- a few seconds over 10 days. You might also measure the drift over the measurement interval and back-calculate the speed at which the clock was running to compensate the times.

If you want to do the calibration a cheap GPS module that outputs 1PPS is a good calibration signal.

It sounds like you might be writing to eeprom. Have you considered writing to flash? That would give you a lot more space and the endurance with the Tiny-2 isn't horrible. You might also want an external eeprom or fram - both are cheap and easy to implement and will give you far more space.

**Reply by ObviousInRetrospect** [reply]
https://lcsc.com/product-detail/Crystals_Seiko-SC-20S-32-768kHz-20PPM-7pF_C97602.html is on the recommended crystal list
and usually takes a 9.5-10pf cap on each side

https://www.digikey.com/en/products/detail/micro-crystal-ag/CM8V-T1A-32-768KHZ-6PF-20PPM-TA-QC/10499094 isn't on the recommended list but seems to work fine if you don't want to wait for China. It tends to want a 10-11pf cap.

p.s. such caps are dirt cheap from lcsc.com (https://lcsc.com/product-detail/Multilayer-Ceramic-Capacitors-MLCC-SMD-SMT_YAGEO-CC0402BRNPO9BN8R2_C327205.html)  and less dirt cheap from digikey https://www.digikey.com/en/products/detail/yageo/CC0402BRNPO9BN8R2/5195037 - I bought basically every value between 7 and 15 I could get my hands on after getting annoyed with the amount of swing a +/- 0.5pf cap adds. 

p.p.s. if you aren't used to 0402 caps you want to assume you will lose 1-5 for every one that actually makes it onto a board.

p.p.p.s its worth slightly expanding the pads for the crystal and not using a footprint that assumes solder paste + a stencil. If you are stuck with undersized pads use a knife tip and slide the crystal onto the pads with both pads having a bit of molten solder. 

p.p.p.p.s.  microchip f*cked up the io multiplexing table even by their usual abysmal standards on the ATtiny-2. Don't bother trying to get 2 UARTS + SPI + XTAL it can't be done on any ATtiny-2.

**Reply by SpenceKonde** [reply]
With a crystal, that's probably viable with care. Without one you have not a prayer. 

OIR: 
Is there a reason you say 0402? I usually use 0805 or maybe 0603, even for crystal loading caps. 0402 is not pleasant to work with. 

As for the pinouts, the original sin was committed on the 0/1-series (the genesis of most of the tinyAVR pinout mistakes) - putting the crystal on the same pins as the uart? What numbskull thought that was a good idea?.It's unsurprising that the SPI and USART use the same pins though - silicon is almost certainly being reused. While in the classic AVR days the tight association was between SPI and I2C, which on tinyAVRs was usually not even separate - they both had to be implemented with a single halfassed thing that they called a USI, which initially sounds like a feature, until you actually read how little it does, the sort of thing that, as you go about implementing the functionality, you're left with a nagging question in your mind: "Is this really better than a software implementation?" (the answer is "yes, but only decisively so in slave mode"). Now, on modern AVRs, the I2C's are alone and the UART and SPI pins overlap. It isn't hard for me to imagine that the implementation might involve shared silicon, as serial and SPI have a ton in common (particularly in sync mode). Maybe the shift registers have to be paired with hardware pins or something for technical reasons (which would explain the almost universal affinity of MOSI and TX, MISO, and RX, etc), and they've likely designed one uniform dual shift register that they stamp onto every mux option of serial or SPI, which all have 4 pins, almost universally in the same order, and usually overlapping exactly with the pins of each other. They do it on the DX parts too. I2C seems to be a separate beast now (which is understandable, since the modern I2C hardware is tightly bound to the protocol, doing a bunch of things for you).  So it's not a surprise that the SPI and USART share pins, the problem is that some damned fool decided to put the external crystal on the same pins as the serial's default TX/RX. But as you may have noticed, they have been sticking to their pinouts unlike ever before. The Tinyies get one pinout (stretching from 14 to 24 pins) , and the non-tinies get a different one (stretching from 14 to 64 pins). They even numbered the ZCD on the DD ZCD3 - why? Because it was on a different pin from ZCD 0-2 because the smallest DD's don't have any of the ZCD input pins and they didn't want to say "Oh on these parts ZCD0 is over here..." so they just gave them a ZCD3 (notice that it's address in memory is that of ZCD0 - ie, it's implemented as ZCD0, but they named it ZCD3 in docs and headers because it's input pin is PC2 instead of PD1, and the input pin of a ZCD isn't muxable so they couldn't even add portmux options), and they add mux options, but don't  remove them (though it's unclear whether the EA will be getting the DD's SPI mux options - if not it would not set an encouraging precedent, I'm liking the consistent pinouts. Product brief says yes, headers say no.  - but the headers and brief agree that it will get the new USART0 pinsets that the DD has). The above shared hardware theory is predicts that where the USART0 mappings go, the SPI0 mappings will also go; I predict the headers are just not done, and before or slightly after product ships, there will be a new ATpack release that adds the new SPI mappings to the EAs.

**Reply by ObviousInRetrospect** [reply]
>Is there a reason you say 0402? I usually use 0805 or maybe 0603, even for crystal loading caps. 0402 is not pleasant to work with.

Few reasons, none of which are great but in aggregate push me that way.

I suspect the core reason is my eyes are generally happier soldering even reasonable stuff under a microscope. At which point it makes a lot less difference.

I've been told by people with more hardware experience that crystals are flakey and the better properties and tighter layout of 0402 is worthwhile for crystal loading caps. I can't say I have done A/B testing. I can say I have seen boards have their oscillators not work well for reasons I don't understand. Years ago when I mostly used 1206 elsewhere (I assume that's what most people start SMT with) I used 0805 on the crystal (atmega328 and similar). I've tended to continue using a size smaller for crystal caps than my default.

I also find it makes the layout easier and I tend to be annoyed at the size of the crystal and the area it blocks from other signal routing (right next to the mcu, premium real-estate, ground plane under, no fast signals allowed in the vicinity). Often when I design a board I add pads for a crystal that I only infrequently populate and so the tradeoff of it taking a couple extra seconds to populate is worth making it smaller.

I mostly use 0603 for everything else (smallest size where resistors are usually labeled) these days. I only find 0402 modestly more annoying. They are prone to tombstoning if you place them and then hit with the iron+solder but if you tin one pad then hold it down and reflow one side its pretty painless. In loading cap sizes you can get good dielectrics in 0402 with plenty of voltage and they are sufficiently cheaper (0.4c vs 2c) that the fact that you lose half of them doesn't matter. I mostly use a bent conical tip which makes 0603 less annoying than 0805 because an 0603 is small enough that the bent conical works a lot like a knife for reworking 0603s but 0805s are a bit too big to easily reflow both sides at once. Bigger caps 0402 starts to push you to worse dielectrics.

I also tend to rarely use caps in the single to low double digit pf range for anything else and so those I stock in 0402. I'm in the process of transitioning my 0.1uf decoupling caps to 0402 for similar reasons since there is no text on them anyway.

Also I have the bad habit of using oshpark for some early test and prototype circuits and that tends to aggravate my tendency to make boards smaller than they should be.

But yeah if you aren't comfortable with SMT in general probably no harm in moving up to 0603 or even 0805. I wouldn't be caught dead using TH or 1206 though for this.
